### PR TITLE
[plugin.program.mceremote] 3.0.0

### PR DESCRIPTION
**Pull request by jhsrennie**
### Description
MCERemote: This addon allows you to configure a Microsoft MCE remote, or any compatible remote using the eHome driver.
This is the port to Python 3 for Matrix.
### Checklist:

- [x] My code follows the [add-on rules](http://kodi.wiki/view/Add-on_rules) and [piracy stance](http://kodi.wiki/view/Official:Forum_rules#Piracy_Policy) of this project. 
- [x] I have read the [CONTRIBUTING](https://github.com/xbmc/repo-plugins/blob/master/CONTRIBUTING.md) document
- [x] Each add-on submission should be a single commit with using the following style: [plugin.video.foo] v1.0.0

Additional information :
- Submitting your add-on to this specific branch makes it available to any Kodi version equal or higher than the branch name with the applicable Kodi dependencies limits.
- [add-on development](http://kodi.wiki/view/Add-on_development) wiki page.
- Kodi [pydocs](http://kodi.wiki/view/PyDocs) provide information about the Python API
- [PEP8](https://www.python.org/dev/peps/pep-0008/) codingstyle which is considered best practice but not mandatory.
- This add-on repository has automated code guideline check which could help you improve your coding. You can find the results of these check at [Codacy](https://www.codacy.com/app/Kodi/repo-plugins/dashboard). You can create your own account as well to continuously monitor your python coding before submitting to repo.
- Development questions can be asked in the [add-on development](http://forum.kodi.tv/forumdisplay.php?fid=26) section on the Kodi forum.
- If you see no activity on your PR after a week (so at least one weekend has passed) then please go to the #kodi-dev freenode IRC channel to reach out to the team

### CODE REVIEW

**Comment by enen92** [comment]
@jhsrennie please look at the errors from addon-checker: https://github.com/xbmc/repo-plugins/pull/3466/checks?check_run_id=2152274446

**Comment by jhsrennie** [comment]
The last change was because after reading other pull request comments I realised I had left the language tag empty. I've now set it to language en /language

**Comment by jhsrennie** [comment]
And I have now split the original code into two files to keep the entry point < 15 lines.

**Comment by jhsrennie** [comment]
The checker is warning of extra files in the resources/data directory. Those files are registry files the user can use to manually set registry entries. I guess they could be removed if this is a problem ...

**Comment by enen92** [comment]
As long as they are not binary files, not executable and not automatically processed by the add-on (use has to do the process manually) it's fine.

File: plugin.program.mceremote/ReadMeFirst.txt [hunk 1]
```diff
@@ -0,0 +1,132 @@
+MCERemote Addon for XBMC
```
Comment (enen92) [inline plugin.program.mceremote/ReadMeFirst.txt]:
```suggestion
MCERemote Addon for Kodi
```

File: plugin.program.mceremote/ReadMeFirst.txt [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+========================
+
+Introduction
+------------
+This plugin will customise the Microsoft remote (and compatibles like
+the Asrock and HP remotes) for use with XBMC. If you don't have an MS
```
Comment (enen92) [inline plugin.program.mceremote/ReadMeFirst.txt]:
```suggestion
the Asrock and HP remotes) for use with Kodi. If you don't have an MS
```

File: plugin.program.mceremote/ReadMeFirst.txt [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+or compatible remote the plugin will warn you of this. If your remote
+isn't MS compatible the addon won't do anything useful, but won't do
+any harm either.
+
+The Microsoft remote works pretty well in XBMC without any
```
Comment (enen92) [inline plugin.program.mceremote/ReadMeFirst.txt]:
```suggestion
The Microsoft remote works pretty well in Kodi without any
```

File: plugin.program.mceremote/ReadMeFirst.txt [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+customisation needed. Just a few buttons won't work. If you're happy
+with this you don't need to use this addon and you need read no
+further.
+
+Quick start
+-----------
+Most people will just want to configure the remote to use the standard
+Windows Media Center keyboard shortcuts. In this case just select the
+second item in the list "Apply current settings to remote". Exit XBMC
```
Comment (enen92) [inline plugin.program.mceremote/ReadMeFirst.txt]:
```suggestion
second item in the list "Apply current settings to remote". Exit Kodi
```

File: plugin.program.mceremote/ReadMeFirst.txt [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+and reboot your PC, and all buttons on the remote will now work with
+XBMC. The remote will still work with Windows Media Center as well. If
```
Comment (enen92) [inline plugin.program.mceremote/ReadMeFirst.txt]:
```suggestion
Kodi. The remote will still work with Windows Media Center as well. If
```

File: plugin.program.mceremote/ReadMeFirst.txt [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+
+After you've configured the remote you don't need the addon again and
+you can uninstall it. The only reason to leave the addon installed is
+if you want to tweak the remote config further.
+
+A note on Windows 10, 7 and Vista: these versions of Windows have a
+feature called User Access Control that (very wisely) stops
+applications overwriting key bits of the registry. If UAC is on you may
+get an error message from the plugin to say it cannot write to the
+registry. If this happens try right-clicking on the XBMC shortcut then
```
Comment (enen92) [inline plugin.program.mceremote/ReadMeFirst.txt]:
```suggestion
registry. If this happens try right-clicking on the kodi shortcut then
```

File: plugin.program.mceremote/ReadMeFirst.txt [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+choose "Run as administrator", and the plugin should now be able to
+write to the registry. You only need to do this once. After you've run
+XBMC as administrator and used the plugin you can go back to using XBMC
```
Comment (enen92) [inline plugin.program.mceremote/ReadMeFirst.txt]:
```suggestion
Kodi as administrator and used the plugin you can go back to using kodi
```

File: plugin.program.mceremote/ReadMeFirst.txt [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+  backspace
+  escape
+  tab
+  space
+  insert
+  home
+  delete
+  end
+
+When you close the Settings dialog be sure to click OK or XBMC will
```
Comment (enen92) [inline plugin.program.mceremote/ReadMeFirst.txt]:
```suggestion
When you close the Settings dialog be sure to click OK or Kodi will
```

File: plugin.program.mceremote/addon.py [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+import mceremote
+
+# **********************************************************************
+# Start execution
+# ---------------
+# **********************************************************************
+
+# Check if we are running on Windows and display an error if not
+if sys.platform != "win32":
+    xbmcgui.Dialog().ok("MCERemote", "This plugin only runs on Windows")
```
Comment (enen92) [inline plugin.program.mceremote/addon.py]: Strings shown on the UI have to be localized. Please add to strings.po

File: plugin.program.mceremote/changelog.txt [hunk 10]
```diff
@@ -0,0 +1,34 @@
+[B]Version 3.0.0 16th March 2021[/B]
```
Comment (enen92) [inline plugin.program.mceremote/changelog.txt:1]: Changelog is no longer displayed in Kodi. You might want to add the most recent changes to the news tag in addon.xml

File: plugin.program.mceremote/mceremote.py [hunk 11]
```diff
@@ -0,0 +1,733 @@
...
+
+        # if the user has entered "mce" but the button table contains no
+        # default, set the keytext to "" so it will be ignored.
+        if keytext == "mce" and len(button[3]) == 0:
+            keytext = ""
+
+        if keytext != "":
+            thisbutton = ConvertKeyText(keytext, button[3])
+            if len(thisbutton) != 3:
+                dialog.ok("MCERemote", "The setting for " + button[0] + " is invalid: " + keytext)
```
Comment (enen92) [inline plugin.program.mceremote/mceremote.py]: Localize

File: plugin.program.mceremote/mceremote.py [hunk 12]
```diff
@@ -0,0 +1,733 @@
...
+            reportmappingtable.append(0)
+            for i in thisbutton:
+                reportmappingtable.append(i)
+
+# Open the registry key
+
+    try:
+        hkey = winreg.OpenKeyEx(winreg.HKEY_LOCAL_MACHINE, _ReportMappingTable, 0, winreg.KEY_SET_VALUE)
+    except WindowsError as e:
+        dialog.ok("MCERemote", "Error opening the IR registry key: " + e.args[1] + ". Try running XBMC as administrator")
```
Comment (enen92) [inline plugin.program.mceremote/mceremote.py]: Localize

File: plugin.program.mceremote/mceremote.py [hunk 13]
```diff
@@ -0,0 +1,733 @@
...
+        return
+    except:
+        dialog.ok("MCERemote", "Unexpected error opening the IR registry key")
```
Comment (enen92) [inline plugin.program.mceremote/mceremote.py]: Localize

File: plugin.program.mceremote/mceremote.py [hunk 14]
```diff
@@ -0,0 +1,733 @@
...
+        return
+
+# Write the ReportMappingTable data
+
+    try:
+        winreg.SetValueEx(hkey, "ReportMappingTable", 0, winreg.REG_BINARY, bytes(reportmappingtable))
+    except Exception as e:
+        dialog.ok("MCERemote", "Unexpected error writing the ReportMappingTable data: " + e.args[0])
```
Comment (enen92) [inline plugin.program.mceremote/mceremote.py]: Localize

File: plugin.program.mceremote/mceremote.py [hunk 15]
```diff
@@ -0,0 +1,733 @@
...
+    except:
+        dialog.ok("MCERemote", "Unexpected error writing the ReportMappingTable data")
```
Comment (enen92) [inline plugin.program.mceremote/mceremote.py]: Localize

File: plugin.program.mceremote/mceremote.py [hunk 16]
```diff
@@ -0,0 +1,733 @@
...
+
+# Finished!
+
+    winreg.CloseKey(hkey)
+    dialog.ok("MCERemote", "Done! Please restart Windows to activate the changes.")
```
Comment (enen92) [inline plugin.program.mceremote/mceremote.py]: Localize

File: plugin.program.mceremote/mceremote.py [hunk 17]
```diff
@@ -0,0 +1,733 @@
...
+# Restore the default Windows ReportMappingTable.
+# **********************************************************************
+def ApplyDefaultSettings(Prompt):
+
+# Require confirmation from the user
+
+    dialog = xbmcgui.Dialog()
+
+    if Prompt:
+        if not dialog.yesno("MCERemote", "This will reset the remote to the default MS config. Do you want to proceed?"):
```
Comment (enen92) [inline plugin.program.mceremote/mceremote.py]: Localize

File: plugin.program.mceremote/mceremote.py [hunk 18]
```diff
@@ -0,0 +1,733 @@
...
+            return
+
+# Open the registry key
+
+    try:
+        hkey = winreg.OpenKeyEx(winreg.HKEY_LOCAL_MACHINE, _ReportMappingTable, 0, winreg.KEY_SET_VALUE)
+    except WindowsError as e:
+        dialog.ok("MCERemote", "Error opening the IR registry key: " + e.args[1] + ". Try running XBMC as administrator")
```
Comment (enen92) [inline plugin.program.mceremote/mceremote.py]: Localize

File: plugin.program.mceremote/mceremote.py [hunk 19]
```diff
@@ -0,0 +1,733 @@
...
+        return
+    except:
+        dialog.ok("MCERemote", "Unexpected error opening the IR registry key")
```
Comment (enen92) [inline plugin.program.mceremote/mceremote.py]: Localize

File: plugin.program.mceremote/mceremote.py [hunk 20]
```diff
@@ -0,0 +1,733 @@
...
+            reportmappingtable.append(0)
+            for i in button[3]:
+                reportmappingtable.append(i)
+
+# Write the ReportMappingTable data
+
+    try:
+        winreg.SetValueEx(hkey, "ReportMappingTable", 0, winreg.REG_BINARY, bytes(reportmappingtable))
+    except Exception as e:
+        dialog.ok("MCERemote", "Unexpected error writing the ReportMappingTable data: " + e.args[0])
```
Comment (enen92) [inline plugin.program.mceremote/mceremote.py]: Localize

File: plugin.program.mceremote/mceremote.py [hunk 21]
```diff
@@ -0,0 +1,733 @@
...
+    except:
+        dialog.ok("MCERemote", "Unexpected error writing the ReportMappingTable data")
```
Comment (enen92) [inline plugin.program.mceremote/mceremote.py]: Localize

File: plugin.program.mceremote/mceremote.py [hunk 22]
```diff
@@ -0,0 +1,733 @@
...
+
+# Finished!
+
+    winreg.CloseKey(hkey)
+    dialog.ok("MCERemote", "Done! Please restart Windows to activate the changes.")
```
Comment (enen92) [inline plugin.program.mceremote/mceremote.py]: Localize

File: plugin.program.mceremote/mceremote.py [hunk 23]
```diff
@@ -0,0 +1,733 @@
...
+# **********************************************************************
+def ConfigureButtonSettings():
+
+# Open the settings dialog
+
+    _settings.openSettings()
+
+# Ask the user if they want to apply these settings
+
+    if xbmcgui.Dialog().yesno("MCERemote", "Do you want to apply these settings to the remote now?"):
```
Comment (enen92) [inline plugin.program.mceremote/mceremote.py]: Localize

File: plugin.program.mceremote/mceremote.py [hunk 24]
```diff
@@ -0,0 +1,733 @@
...
+# RestoreDefaultButtonSettings
+# ----------------------------
+# Restore the default settings
+# **********************************************************************
+def RestoreDefaultButtonSettings():
+
+# Require confirmation from the user
+
+    dialog = xbmcgui.Dialog()
+    if not dialog.yesno("MCERemote", "This will reset the remote button settings to the Media Center defaults. Do you want to proceed?"):
```
Comment (enen92) [inline plugin.program.mceremote/mceremote.py]: Localize

File: plugin.program.mceremote/mceremote.py [hunk 25]
```diff
@@ -0,0 +1,733 @@
...
+        return
+
+# Restore the default settings from the info in _buttonData
+
+    for nextbutton in _buttonData:
+        _settings.setSetting(nextbutton[0], nextbutton[2])
+
+# Ask the user if they want to apply these settings
+
+    if xbmcgui.Dialog().yesno("MCERemote", "Do you want to apply these settings to the remote now?"):
```
Comment (enen92) [inline plugin.program.mceremote/mceremote.py]: Localize

File: plugin.program.mceremote/mceremote.py [hunk 26]
```diff
@@ -0,0 +1,733 @@
...
+    import os
+    import subprocess
+
+    dialog = xbmcgui.Dialog()
+
+# Check if we are updating KeyMapEdit.exe
+
+    doupdate = _settings.getSetting("update_keyedit")
+    if doupdate == "true":
+        if dialog.yesno("MCERemote", "Do you want to update the keymap editor from Sourceforge?"):
```
Comment (enen92) [inline plugin.program.mceremote/mceremote.py]: You can't update keymaps from remote locations.

File: plugin.program.mceremote/mceremote.py [hunk 27]
```diff
@@ -0,0 +1,733 @@
...
+            if GetKeyMapEdit():
+                _settings.setSetting("update_keyedit", "false")
+            else:
+                return
+
+# Check whether keyboard.xml exists
+
+    dstpath = xbmcvfs.translatePath("special://home/userdata/keymaps/keyboard.xml")
+    if not os.path.isfile(dstpath):
+        if dialog.yesno("MCERemote", "You don't currently have a keyboard.xml file. Do you want to create one?"):
```
Comment (enen92) [inline plugin.program.mceremote/mceremote.py]: Localize

File: plugin.program.mceremote/mceremote.py [hunk 28]
```diff
@@ -0,0 +1,733 @@
...
+            CreateKeyboardDotXML()
+        else:
+            return
+
+# Select the keymap editor: if KeyMapEdit.exe exists use it, otherwise
+# use Notepad.
+
+    srcpath = _DataDir + "KeyMapEdit.exe"
```
Comment (enen92) [inline plugin.program.mceremote/mceremote.py]: What is this executable?

**Review by enen92 — CHANGES_REQUESTED** [review]
I am not sure we can accept this in the repo as it modifies the XML files directly. I'll request the review from someone else.

File: plugin.program.mceremote/mceremote.py (hunk 26 above)
Comment (jhsrennie) [reply plugin.program.mceremote/mceremote.py]: It does not update the keymap from a remote location. If no keyboard.xml exists in the user's userdata directory it copies a template from the resources/data folder. The only thing it downloads is my keymap editor applet. I guess I can remove the option to download the applet, but then users have to either hand edit the keymap in a text editor or manually download and copy the applet.

File: plugin.program.mceremote/mceremote.py (hunk 28 above)
Comment (jhsrennie) [reply plugin.program.mceremote/mceremote.py]: This is my keymap editor applet. The same applet that MCERemote has been using for all editions before Matrix for the last ten years.

**Comment by jhsrennie** [comment]
Thanks, I will change all references to XBMC to Kodi and add the strings to the po file.

**Comment by jhsrennie** [comment]
On reflection I can see why you'd be nervous about an addon downloading binaries from a site you do not control. I would be happy to remove this code if you want.

Note that this code is in the current (pre-Matrix) version of the addon so you might want to disable it until I can remove the code from the older version as well.

**Comment by enen92** [comment]
Yeah I'd prefer you remove the code and just provide instructions for users to download it if they wish in the readme file or forum post. This is a no go unfortunately

**Comment by jhsrennie** [comment]
Right, I've localised all the strings and removed the download. I think this addresses all the points raised.

File: plugin.program.mceremote/addon.xml [hunk 29]
```diff
@@ -0,0 +1,24 @@
...
+       provider-name="John Rennie">
+  <requires>
+    <import addon="xbmc.python" version="3.0.0"/>
+  </requires>
+  <extension point="xbmc.python.pluginsource" library="addon.py">
+    <provides>executable</provides>
+  </extension>
+  <extension point="xbmc.addon.metadata">
+    <platform>windx</platform>
+    <language>en</language>
```
Comment (enen92) [inline plugin.program.mceremote/addon.xml]: this does not apply, the addon does not provide content (video, music, etc)

**Review by enen92 — CHANGES_REQUESTED** [review]
Minor. This addon only applies to windows and the user is always prompted so It's fine to me.

File: plugin.program.mceremote/addon.xml (hunk 29 above)
Comment (jhsrennie) [reply plugin.program.mceremote/addon.xml]: Can you clarify what the problem is? Is it the platform tag that should be changed?

Comment (enen92) [reply plugin.program.mceremote/addon.xml]: The language one, it refers to content. Your add-on just configures a remote

Comment (enen92) [reply plugin.program.mceremote/addon.xml]: Please remove it

Comment (jhsrennie) [reply plugin.program.mceremote/addon.xml]: Done!

**Review by enen92 — APPROVED** [review]

**Pull request merged**